src/backend/commands/tablespace.c

```diff
@@ -1680,22 +1689,23 @@ PrepareTempTablespaces(void)
 		return;
 
 	/* Need a modifiable copy of string */
-	rawname = pstrdup(temp_tablespaces);
+	rawname = pstrdup(str);
 
 	/* Parse string into list of identifiers */
 	if (!SplitIdentifierString(rawname, ',', &namelist))
 	{
 		/* syntax error in name list */
-		SetTempTablespaces(NULL, 0);
+		*tblSpcs = NULL;
+		*numSpcs = 0;
 		pfree(rawname);
 		list_free(namelist);
 		return;
 	}
 
 	/* Store tablespace OIDs in an array in TopTransactionContext */
-	tblSpcs = (Oid *) MemoryContextAlloc(TopTransactionContext,
+	iTblSpcs = (Oid *) MemoryContextAlloc(TopTransactionContext,
```

Comment:
```suggestion
	*tblSpcs = (Oid *) MemoryContextAlloc(TopTransactionContext,
```

Reply:
Why? We keep internal `tblSpcs` (`iTblSpcs`) and assign them to the original values when we are done. If we fail in the beginning, the original values are set to 0.

Reply:
> If we fail in the beginning, the original values are set to 0.

How/why?

Reply:
If we [fail to split the string to a list](https://github.com/arenadata/gpdb/blob/95b5d8db5bd61354350093eec04b459da2a92aa0/src/backend/commands/tablespace.c#L1695-L1703), we set the original values to 0. 0 (instead of -1) in `numSpcs` marks that there were no tablespaces specified, and that this GUC was already parsed. Otherwise, we fill the internal values and assign them to original ones

Reply:
> If we [fail to split the string to a list](https://github.com/arenadata/gpdb/blob/95b5d8db5bd61354350093eec04b459da2a92aa0/src/backend/commands/tablespace.c#L1695-L1703), we

will not reach the [line](https://github.com/arenadata/gpdb/pull/680/files#diff-f2c91c95b7f2a84d916138e0af4338859803a03cee0d7e2e710fbcb869c59d0cR1706) of code I specified

Reply:
I don't see a problem?

Reply:
I suggest avoid using local vars i*, instead use argument vars

Reply:
Local variables are introduced to avoid writing `(*tblSpcs)[(*numSpcs)++] = InvalidOid;`, instead we can just write `iTblSpcs[iNumSpcs++] = InvalidOid;` and assign them in the end.

Reply:
> (*tblSpcs)[(*numSpcs)++] = InvalidOid;

braces are redundant
```c
*tblSpcs[*numSpcs++] = InvalidOid;
```

Reply:
Without parentheses, `[]` has higher precedence than `*`, so we will index into the incorrect array, and increment the `numSpcs` pointer, not it's value

Reply:
ok, there is [example](https://github.com/arenadata/gpdb/blob/f6f95b6622a3fbdf441ec86164d5921b472694d7/src/backend/optimizer/plan/plangroupext.c#L524) in codebase